### PR TITLE
BoostTrack 

### PR DESCRIPTION
**Pull request by Fleyderer**
Hello! I've decided to add BoostTrack to boxmot, because I am currently researching this tracker results and can't wait for someone else to do it :smiley: 

There are many comments, and I could forget something important:

###  1. Reproduce BoostTrack tracker results

Originally, after running BoostTrack scripts from original repository with `val_half.json`, i got:

```
{"HOTA": 68.486, "MOTA": 75.508, "IDF1": 81.277}
```

To reproduce these results in boxmot:

1)  Download [val_half](https://drive.google.com/file/d/1QZXuLZq6zp4_DrOB0kCwYkclNEksFBEh/view?usp=drive_link) from my drive or prepare it by yourself (It contains already splitted MOT17 validation part). Folder path should be `val_utils/data/MOT17/val_half`
2) Download ByteTrack ablation model from [here](https://drive.google.com/file/d/1iqhM-6V_r1FpOlOzrdP_Ejshgk0DxOob/view), move it to boxmot folder and rename it to `yolox_x_ablation.pth`
2) run command:

```
python3 tracking/val.py --yolo-model yolox_x_ablation.pth --reid-model osnet_x1_0_msmt17.pt --tracking-method boosttrack --verbose --source ./tracking/val_utils/data/MOT17/val_half/ --conf 0.1
```

Results should be:

```
{"HOTA": 68.382, "MOTA": 75.558, "IDF1": 81.154}`
```

For closest results you need to add these lines to [val.py](https://github.com/mikel-brostrom/boxmot/blob/dbce5f2bc4eaecbdce702c600ed0cc04fbe6e43b/tracking/val.py#L277) (between `embs = frame_dets_n_embs[:, 7:]` and `tracker.update`):

```
if args.tracking_method == "boosttrack":
            with open(source.parent / "seqinfo.ini") as f: 
                seq_data = f.readlines()

            x = next(sd for sd in seq_data if sd.startswith("frameRate"))
            frame_rate = int(x[x.find("=") + 1:x.find('\n')])

            tracker.max_age = max(30, frame_rate * 2)
```

Idea: authors of boosttrack added [dynamic](https://github.com/vukasin-stanojevic/BoostTrack/blob/8e80cd263ec33144db91c356d194bfcdaa9894bf/default_settings.py#L64) `max_age` a.k.a. `max_time_lost` to their tracker, which uses `max(30, frame_rate * 2)` of each MOT17, MOT20 sequence.

After this change results should be:

```
{"HOTA": 68.517, "MOTA": 75.55, "IDF1": 81.398}
```

I could also reproduce the original result one to one, but it requires too many steps to show it here. The most important thing is that the tracker works exactly the same as the original, the only difference here is in the processing of detections, which is a bit different in boxmot and boosttrack original repo.

### 2. Comparison: BoostTrack vs ByteTrack

In my experience, this tracker shows speed like 70-75 FPS, while bytetrack works on 200 FPS. Main problem is of course in ECC. I turned it off to check:

BoostTrack without ecc: `{"HOTA": 67.42, "MOTA": 75.01, "IDF1": 79.34}`, and FPS is almost equal to ByteTrack
ByteTrack: `{"HOTA": 67.619, "MOTA": 78.081, "IDF1": 79.188}`.

So my verdict is that default BoostTrack is essentially the same or even worse than ByteTrack.

### 3. BoostTrack+ and BoostTrack++

1) BoostTrack+  == BoostTrack with ReID. Currently I didn't check this tracker with ReID, I will do it in next iteration.

2) Of course, main contribution of this tracker is BoostTrack++ version, but currently I don't know how to better add both + and ++ to boxmot. There are only few changes in configs, and now there is no functionality in boxmot to add tracker without changing its class while changing only its config file. This thing should be discussed and maybe you have your own opinion.

#1826

### CODE REVIEW

**Comment by Fleyderer** [comment]
I will check BoxMOT CI / tests tomorrow

**Comment by mikel-brostrom** [comment]
Nice @Fleyderer!

Add the model here:

https://github.com/mikel-brostrom/boxmot/blob/dbce5f2bc4eaecbdce702c600ed0cc04fbe6e43b/.github/workflows/benchmark.yml#L28C9-L28C16

for automatic benchmarking 😄

**Comment by Fleyderer** [comment]
In tests, there is only one problem with boosttrack: 

```FAILED tests/unit/test_trackers.py::test_tracker_output_size[boosttrack] - assert (1, 8) == (2, 8)```

This happens because `filter_outputs` function of boosttrack filters out one of tracks, which has aspect_ratio 1.61, while aspect_ratio_thresh in boosttrack is 1.6

I think it is not a problem, because it is original logic of this tracker, and strongsort does not pass this test too.

**Comment by mikel-brostrom** [comment]
Let's change that specific bbox:

[144,212,400,480,0.82,0]

Will do

**Comment by mikel-brostrom** [comment]
Also add the tracking method name here:

https://github.com/mikel-brostrom/boxmot/blob/6486fc79e0b2d066cfe8a02df9054e0af7cc5f21/.github/workflows/ci.yml#L114

**Comment by Fleyderer** [comment]
Well, I think we're done with BoostTrack.

**Comment by mikel-brostrom** [comment]
The set of detections used for each may not be exactly the same. When i run the benchmark job I get the following:

With ECC
``` 
                                      Tracker	Status	HOTA↑	MOTA↑	IDF1↑	FPS
[botsort](https://arxiv.org/abs/2206.14651)	✅	68.774	77.176	80.997	46
[strongsort](https://arxiv.org/abs/2202.13514)	✅	68.642	76.409	81.242	17
[boosttrack]	                                ✅	67.683	75.764	80.181	
[bytetrack](https://arxiv.org/abs/2110.06864)	✅	66.773	76.929	77.874	1265
[ocsort](https://arxiv.org/abs/2203.14360)	✅	65.469	74.934	75.983	1483
[imprassoc]	                                ✅	64.36	76.578	71.903	26
[deepocsort](https://arxiv.org/abs/2302.11813)	✅	63.18	74.474	73.496	12
[hybridsort](https://arxiv.org/abs/2308.00783)	❌
```
Without ECC
```
                                      Tracker	Status	HOTA↑	MOTA↑	IDF1↑	FPS
[botsort](https://arxiv.org/abs/2206.14651)	✅	68.774	77.176	80.997	46
[strongsort](https://arxiv.org/abs/2202.13514)	✅	68.642	76.409	81.242	17
[boosttrack]	                                ✅	66.894	75.413	78.351
[bytetrack](https://arxiv.org/abs/2110.06864)	✅	66.773	76.929	77.874	1265
[ocsort](https://arxiv.org/abs/2203.14360)	✅	65.469	74.934	75.983	1483
[imprassoc]	                                ✅	64.36	76.578	71.903	26
[deepocsort](https://arxiv.org/abs/2302.11813)	✅	63.18	74.474	73.496	12
[hybridsort](https://arxiv.org/abs/2308.00783)	❌
```
A bit better than ByteTrack still

**Comment by Fleyderer** [comment]
With my detections, which gives closest to original boosttrack results, without ecc bytetrack is better. I am a bit tired of these small differences in metrics, which are more a matter of chance than of real advantage :smile:

But what about FPS without ECC? I see that FPS here is the same. 

BTW, what do you think about adding + and ++ versions? I am currently working on it, and I would like to know how it will be included in the boxmot. I think, it would be better to add this tracker as 'boosttrack' and let people decide which parameters they need to use or not. And use the BoostTrack++ config by default (when it will be ready).

**Comment by mikel-brostrom** [comment]
> Of course, main contribution of this tracker is BoostTrack++ version, but currently I don't know how to better add both + and ++ to boxmot. There are only few changes in configs, and now there is no functionality in boxmot to add tracker without changing its class while changing only its config file. This thing should be discussed and maybe you have your own opinion.

My usual way of working regarding this is that the config under the repo represents the maximum tracker performance (++). Maybe a comment in the config is enough with the `regular`, `+` and `++` configurations?

**Comment by mikel-brostrom** [comment]
`I think, it would be better to add this tracker as 'boosttrack' and let people decide which parameters they need to use or not. And use the BoostTrack++ config by default (when it will be ready).`

Exactly 🚀

**Comment by Fleyderer** [comment]
> I think, it would be better to add this tracker as 'boosttrack' and let people decide which parameters they need to use or not. And use the BoostTrack++ config by default (when it will be ready).

> My usual way of working regarding this is that the config under the repo represents the maximum tracker performance (++). Maybe a comment in the config is enough with the `regular`, `+` and `++` configurations

great minds think alike :smile:

**Comment by mikel-brostrom** [comment]
Sorry FPS for without ECC updated 😄

**Comment by mikel-brostrom** [comment]
I merge this when the config is updated then 😄

**Comment by Fleyderer** [comment]
Succesfully added BoostTrack+ and ++ versions.

Comments:

1. Fixed normalization in `base_backend.py` to normalize each vector independently. In most cases, when cos_sim is used, it does not matter, but in boosttrack, where `emb_cost = dets_embs @ tracks_embs` it would be incorrect.
2. Added fix #1793 back, idk why did it gone from master :smile: 

### 3. Reproduce 

Original BoostTrack: `{"HOTA": 68.486, "MOTA": 75.508, "IDF1": 81.277}`

BoostTrack+: `{"HOTA": 69.237, "MOTA": 75.623, "IDF1": 82.696}`
BoostTrack++: `{"HOTA": 69.254, "MOTA": 76.196, "IDF1": 82.844}`

Boxmot BoostTrack  with `osnet_x1_0_msmt17.pt`:
BoostTrack+: `{"HOTA": 69.074, "MOTA": 75.554, "IDF1": 82.033}`
BoostTrack++: `{"HOTA": 69.177, "MOTA": 76.146, "IDF1": 82.485}`

Boxmot BoostTrack with original BoostTrack ReID:
BoostTrack+: `{"HOTA": 69.007, "MOTA": 75.546, "IDF1": 81.936}`
BoostTrack++: `{"HOTA": 69.229, "MOTA": 76.255, "IDF1": 82.333}`

How to add original ReID:

1. Change NR_CLASSES_DICT `reid_model_factory.py`:

```
NR_CLASSES_DICT = {
    'market1501': 751,
    'duke': 702,
    'veri': 576,
    'vehicleid': 576,
    'boost': 2510
}
```

2. Download weights from [here](https://drive.google.com/file/d/1RCV1ZxzZCjGKV-kYyZaTIwG4luoIfa7B/view?usp=drive_link), put it in boxmot directory and rename from `osnet_ain_ms_d_c.pth.tar` to `osnet_ain_x1_0_boost.pt`.

**Comment by Fleyderer** [comment]
One more thing: does the new FPS feature take into account the calculation of embeddings in online mode, and not pre-calculated ones?

**Comment by mikel-brostrom** [comment]
Not pre-calculated. At the moment it is just a print in one of the tests, so it has to be added to the README by hand. It takes the average of 100 updates, not counting the first warmup iteration.

**Comment by mikel-brostrom** [comment]
Hmmm. The benchmark gave very low results

**Comment by mikel-brostrom** [comment]
The rest of the trackers got a small boost

**Comment by Fleyderer** [comment]
> Hmmm. The benchmark gave very low results for boosttrack : https://github.com/mikel-brostrom/boxmot/pull/1849/files. Could you have a look at it?

I am trying to reproduce results step by step:

Firstly, I have found [mot-metrics-benchmark](https://github.com/mikel-brostrom/boxmot/blob/210e4d0f2c1f80c39aa571c1f89b65e897890c22/.github/workflows/benchmark.yml#L21)

Secondly, I've downloaded data using steps from this job:

```
wget https://github.com/mikel-brostrom/boxmot/releases/download/v10.0.83/MOT17-50.zip -O tracking/val_utils/MOT17-50.zip
unzip tracking/val_utils/MOT17-50.zip -d ./tracking/val_utils/data/
```

Thirdly, ran val.py:

```
python3 tracking/val.py --imgsz 320 --classes 0 --yolo-model yolov8x.pt --reid-model osnet_x1_0_dukemtmcreid.pt --tracking-method boosttrack --verbose --source ./tracking/val_utils/data/MOT17-50/train
```

And got some errors:

```
2025-03-07 09:14:38.698 | ERROR    | tracking.utils:download_mot_eval_tools:135 - Error processing /home/kozakmv/Programming/Python/boxmot/tracking/val_utils/data/__MACOSX/MOT17-50/train/MOT17-05/gt/._gt.txt: 'utf-8' codec can't decode byte 0xa9 in position 37: invalid start byte
```

And also results are:

```
{"HOTA": 33.781, "MOTA": 22.394, "IDF1": 35.11}
```

For bytetrack results are:

```
{"HOTA": 33.661, "MOTA": 23.004, "IDF1": 35.442}
```

These results are not even close to HOTA 68+.  I want to reproduce results by myself, without already existing dets_n_embs cache, because there were problems with embeddings normalization etc.

**Comment by mikel-brostrom** [comment]
Benchmark here: https://github.com/mikel-brostrom/boxmot/blob/master/.github/workflows/benchmark.yml 😄

**Comment by Fleyderer** [comment]
> Benchmark here: https://github.com/mikel-brostrom/boxmot/blob/master/.github/workflows/benchmark.yml 😄

Yes, it is the same link I provided. There is only one command with `val.py`:

```
tracking/val.py --imgsz 320 --classes 0 --yolo-model yolov8x.pt --reid-model osnet_x1_0_dukemtmcreid.pt --tracking-method ${{ matrix.tracker }} --verbose --source ./tracking/val_utils/data/MOT17-50/train
```

**Comment by mikel-brostrom** [comment]
It is making use of preexisting detections and embeddings, notice that the job downloads a zip file that contains a folder yolov8x and osnet_x1_0_dukemtmcreid. The detections and embeddings do not come from these models though, but from YOLOX-X and BoT.

**Comment by Fleyderer** [comment]
> It is making use of preexisting detections and embeddings, notice that the job downloads a zip file that contains a folder yolov8x and osnet_x1_0_dukemtmcreid. The detections and embeddings do not come from these models though, but from YOLOX-X and BoT. So it is a bit confusing. This is to reduce the benchmarking time to minimum.

As i said:

> I think we should reproduce results without already existing dets_n_embs cache, because it should be fully reproducible and comparable, and for example there were problems with embeddings normalization, earlier there were problems with original ByteTrack YOLOX image preprocessing etc. 

That's because current preexisting dets, embs may have mistakes.

We could just recalculate them now, when problems are fixed, and use in current and further benchmarks

**Comment by mikel-brostrom** [comment]
> We could just recalculate them now, because problems are fixed, and use in current and further benchmarks

Sounds great! Given that your results are better than with the existing dets and embs 😄

**Comment by Fleyderer** [comment]
And these metrics, which are given in benchmark, should be reproducible by every possible user of boxmot, else these metrics are not valuable.

It's just that when I work with any project and can reproduce the metrics one to one, I can trust this project. In any other case, I can not. :slightly_smiling_face:

**Comment by mikel-brostrom** [comment]
A download of this dets and embs file should suffice to make the results reproducible everywhere

**Comment by Fleyderer** [comment]
> A download of this dets and embs file should suffice to make the results reproducible everywhere

Yes, and I also mean that users should have ability to reproduce results from data, because dets-embs file could be possibly changed by authors, or outdated, or post-processed.

I think, the best way is to have dataset.zip (Like MOT17-50), and recalculate these dets-embs sometimes, like now, because some mistakes could be fixed and dets-embs are changed. And also provide instructions like "wget dataset.zip, unzip, run val.py, profit"

Maybe we will never recalculate them again, but main idea is that we should never fully trust something precalculated. 

P. S. And we still cannot trust MOT17 or MOT20 datasets, because the detections were not collected manually, but with the model... But that's a completely different story :smile:

**Comment by Fleyderer** [comment]
Yes, I can generate folder with the same structure, but is it ok to put data from yolox_x_ablation to yolov8x dets and embs?

yolov8x will not give results near HOTA 68, because it is just pretrained detector, while yolox_x_ablation is a generally accepted model against which detections all metrics are calculated, including MOTChallenge.

I can just put it in yolov8x folder, but I think that best way is to add bytetrack [ablation](https://drive.google.com/file/d/1iqhM-6V_r1FpOlOzrdP_Ejshgk0DxOob/view) model to `YOLOX_ZOO` with name `yolox_x_ablation.pt` and also allow everyone use this reference model without overheads. Besides this, change this folder from yolov8x to yolox_x_ablation, change benchmark.yml commands too.

I can add needed changes, if you agree

**Comment by mikel-brostrom** [comment]
> I think that best way is to add bytetrack [ablation](https://drive.google.com/file/d/1iqhM-6V_r1FpOlOzrdP_Ejshgk0DxOob/view) model to YOLOX_ZOO with name yolox_x_ablation.pt

Good idea!

> yolov8x will not give results near HOTA 68

Yes, this is clear to me. The fact that yolov8x model name is used is just for the sake of loading the predefined dets and embs that end up in folders called like this

**Comment by Fleyderer** [comment]
There is one more thing to discuss: `--conf` value. Different trackers uses different min confidence values, but for example ByteTrack has no hyperparameter for minimum confidence value.

I think, I have to generate detections with all possible confidences, like `--conf 0.0` and add min-conf hyperparam to trackers, if required. Thus, each tracker will be able to use the detections it needs and filter on its own side, and not in advance

One more thing that I see: ByteTrack has better results with lower confidence. We can have a look at another trackers, maybe they will give better HOTA values too.

**Comment by mikel-brostrom** [comment]
> I think, I have to generate detections with all possible confidences, like --conf 0.0 and add min-conf hyperparam to trackers, if required. Thus, each tracker will be able to use the detections it needs and filter on its own side, and not in advance

I like this. It is probably the best way of handling the unique conf value in each tracker. 

> One more thing that I see: ByteTrack has better results with lower confidence. We can have a look at another trackers, maybe they will give better HOTA values too.

If you generate these detections and embeddings we could run an evolve job to figure this out

**Comment by Fleyderer** [comment]
Yes, I did it and while there are many small changes, I will create new pull request

**Comment by mikel-brostrom** [comment]
Looking forward to this 🚀
